cleanlab/dataset.py

```diff
@@ -0,0 +1,150 @@
+    should be merged.
+
+    The method uses the joint distribution of noisy and true labels to compute ontological issues
+    via the approach published in (Northcutt, Jiang, Chuang (2021) JAIR).
+    paper: https://arxiv.org/abs/1911.00068
+
+    This method works by providing any one (and only one) of the following inputs:
+    1. confident_joint
+    2. joint
+    3. labels and pred_probs
+    but it is not necessary to provide a combination, e.g. the labels and pred_probs and joint.
+
+    For parameter info, see the docstring of `count.estimate_joint`
+
+    Parameters
+    ----------
+    See `count.estimate_joint` docstring.
+
+    asymmetric : bool (default: True)
+      If `asymmetric==True`, includes both pairs (class1, class2) and (class2, class1). Use this for
+      finding "is a" relationships where for example "class1 is a class2".
+      If `asymmetric==False`, the pair (class1, class2) will onyl be returned once and order is
+      arbitrary (internally this is just summing score(class1, class2) + score(class2, class1)).
+
+    Returns
+    -------
+        A tuple of 2 ordered lists: (list_of_pairs_of_classes, likelihood_they_should_be_merged)"""
+
+    def _find_largest_pairs(matrix):
+        """Helper function that returns the sorted (descending) (row, column) index pairs
+        and their values for an arbitrary 2-d numpy matrix."""
+
+        # Find indices of similar classes
+        idx = np.array(list(zip(*np.unravel_index(np.argsort(matrix.ravel()), matrix.shape)))[::-1])
+        # Bring max score between 0.1 and 1.0, otherwise score scales inversely with num_classes ^ 2
+        scaling_factor = 10 ** np.floor(-np.log10(np.max(matrix)))
+        scores = np.sort(matrix.ravel()) * scaling_factor
+        return idx, scores
+
+    if joint is None:
+        joint = estimate_joint(
+            labels=labels,
+            pred_probs=pred_probs,
+            confident_joint=confident_joint,
+            multi_label=multi_label,
+        )
+    if asymmetric:
+        # Remove diagonal elements
+        joint_no_diagonal = joint - np.eye(len(joint)) * joint.diagonal()
+        idx, scores = _find_largest_pairs(joint_no_diagonal)
+        cutoff = len(joint) ** 2 - len(joint)  # Omit diagonal elements from end of ranking
+        return idx[:cutoff], scores[:cutoff]
+    else:  # symmetric
+        # Sum the upper and lower triangles and remove the lower triangle and the diagonal
+        # This provides values only in (the upper triangle) of the matrix.
+        sym_joint = np.triu(joint) + np.tril(joint).T
+        sym_joint = sym_joint - np.eye(len(sym_joint)) * sym_joint.diagonal()
+        idx, scores = _find_largest_pairs(sym_joint)
+        cutoff = (len(joint) ** 2 - len(joint)) / 2  # Omit lower triangle and diagonal elements
+        return idx[:cutoff], scores[:cutoff]
+
+
+def get_classes_ranked_by_label_noise(joint, descending=True, rank_by="given_label"):
```

Comment:
All my comments below for `get_classes_ranked_by_label_quality` apply to this method as well